### PR TITLE
Allow access using a Jenkins API token without an OIDC Session

### PR DESCRIPTION
**Pull request by mikecirioli**
Proposing a new option of the `OicSecurityRealm` that will allow an administrator to Allow access using a Jenkins API token without an OIDC Session.

Currently, when using the `oic-auth` plugin, if a user explicitly logs out of the IdP then they will now longer be able to perform any sort of action using `jenkins api tokens` because the `oic-auth` plugin will reject the request due to the `OicSession` being null.  This behavior may make sense for some users, but it can cause problems for other users who may be relying on external automation tools which make use of `jenkins api tokens` for authentication.

This PR introduces a new (optional, disabled by default) configuration which lets an admin re-enable the traditional jenkins api token access behavior.  This is accomplished by checking to see if a request appears to be using a valid `jenkins api token`, and if so hands of the request processing to the next filter in the chain instead of procssessing it as an OIC based access request.

image

### Testing done
new unit test has been added to validate the new functionality

### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/main/resources/org/jenkinsci/plugins/oic/OicSecurityRealm/config.properties [hunk 1]
```diff
@@ -38,4 +38,5 @@ UsernameFieldName=User name field name
 WellknownConfigurationEndpoint=Well-known configuration endpoint
 UseRefreshTokens=Enable Token Refresh using Refresh Tokens
 DisableTokenExpirationCheck=Disable Token Expiration Check
-AllowedTokenExpirationClockSkewSeconds=Token Expiry Expiration Clock Skew
\ No newline at end of file
+AllowedTokenExpirationClockSkewSeconds=Token Expiry Expiration Clock Skew
+EnableTraditionalApiTokenAccess=Enable Tradition Jenkins API token access
```
Comment (amuniz) [inline src/main/resources/org/jenkinsci/plugins/oic/OicSecurityRealm/config.properties:42]:
```suggestion
EnableTraditionalApiTokenAccess=Enable Traditional Jenkins API token access
```

File: src/main/resources/org/jenkinsci/plugins/oic/OicSecurityRealm/help-traditionalApiTokenAccessEnabled.html [hunk 2]
```diff
@@ -0,0 +1,7 @@
+<div>
+    Enabling this functionality allows Jenkins API token based access even if the associated user has
+    completly logged out from Jenkins and the OIC Provider.
+
+    The default behavior is to require any Jenkins API token based access to have an valid OIC session user
+    session associated with it.
```
Comment (amuniz) [inline src/main/resources/org/jenkinsci/plugins/oic/OicSecurityRealm/help-traditionalApiTokenAccessEnabled.html:6]: I would explicitly mention that the user needs to be logged-in in the UI to be able to use an API Token from the CLI or HTTP API.

File: src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java [hunk 3]
```diff
@@ -1394,6 +1409,25 @@ public boolean handleTokenExpiration(HttpServletRequest httpRequest, HttpServlet
...
+        if (isTraditionalApiTokenAccessEnabled()) {
+            // check if this is a valid api token based request
+            if (httpRequest.getHeader("Authorization") != null) {
+                String token = new String(
+                                Base64.getDecoder()
+                                        .decode(httpRequest
+                                                .getHeader("Authorization")
+                                                .substring(6)),
+                                StandardCharsets.UTF_8)
+                        .split(":")[1];
```
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java:1421]: Isn't there any utility in core to do this?

File: src/test/java/org/jenkinsci/plugins/oic/PluginTest.java [hunk 4]
```diff
@@ -943,6 +977,52 @@ public void loginWithCheckTokenFailure() throws Exception {
...
+
+        // create a jenkins api token for the test user
+        String token = User.getById(TEST_USER_USERNAME, false)
+                .getProperty(ApiTokenProperty.class)
+                .generateNewToken("foo")
+                .plainValue;
+
+        // validate that the token can be used
+        HttpResponse<String> rsp = getPageWithGet(TEST_USER_USERNAME, token, "/whoAmI/api/xml");
+        MatcherAssert.assertThat("response should have been 200\n" + rsp.body(), rsp.statusCode(), is(200));
```
Comment (amuniz) [inline src/test/java/org/jenkinsci/plugins/oic/PluginTest.java:999]: I assume this fails if `isTraditionalApiTokenAccessEnabled == false`,  right?

Comment (mikecirioli) [reply src/test/java/org/jenkinsci/plugins/oic/PluginTest.java:999]: yes, that is what [this next bit](https://github.com/jenkinsci/oic-auth-plugin/pull/386/files/9dafa6a67326e37371c4583b187de3fbc7216074#diff-fbf2979fa131a62f8efe7b73f007ba2048436733e151acf28b80836a70a7e901R1009) of the test checks

File: src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java (hunk 3 above)
Comment (mikecirioli) [reply src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java:1421]: did not find anything with a mild search, happy to change it if you can point me at the code.

File: src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java [hunk 5]
```diff
@@ -1394,6 +1409,25 @@ public boolean handleTokenExpiration(HttpServletRequest httpRequest, HttpServlet
 
         User user = User.get2(authentication);
 
+        if (isTraditionalApiTokenAccessEnabled()) {
+            // check if this is a valid api token based request
+            if (httpRequest.getHeader("Authorization") != null) {
+                String token = new String(
+                                Base64.getDecoder()
+                                        .decode(httpRequest
+                                                .getHeader("Authorization")
```
Comment (Vlatombe) [inline src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java:1418]: Should check the expected format, you never now.
```suggestion
            String authHeader = httpRequest.getHeader("Authorization");
            if (authHeader != null && authHeader.startsWith("Basic ")) {
                String token = new String(
                                Base64.getDecoder()
                                        .decode(authHeader
```

**Review by amuniz — APPROVED** [review]
Looks good to me (needs spotless fixes).

File: src/main/resources/org/jenkinsci/plugins/oic/OicSecurityRealm/config.properties [hunk 6]
```diff
@@ -38,4 +38,5 @@ UsernameFieldName=User name field name
 WellknownConfigurationEndpoint=Well-known configuration endpoint
 UseRefreshTokens=Enable Token Refresh using Refresh Tokens
 DisableTokenExpirationCheck=Disable Token Expiration Check
-AllowedTokenExpirationClockSkewSeconds=Token Expiry Expiration Clock Skew
\ No newline at end of file
+AllowedTokenExpirationClockSkewSeconds=Token Expiry Expiration Clock Skew
+EnableTraditionalApiTokenAccess=Enable Traditional Jenkins API token access
```
Comment (jtnord) [inline src/main/resources/org/jenkinsci/plugins/oic/OicSecurityRealm/config.properties:42]: NIT: this to me is confusing, you are not enabling traditional Jenkins API token access here, you are mearly allowing its usage without a current OIdC Session.
I would suggest the name and text it changed to be something like `allowTokenAccessWIthoutSession` and "Allow Access with a Jenkins API token without an OIDC Session"

**Comment by mikecirioli** [comment]
I plan to merge this at the end of the day today unless anyone has an objection

cc: @jenkinsci/oic-auth-plugin-developers

**Pull request merged**